jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java

```diff
@@ -171,36 +319,111 @@ public synchronized void setScanInterval(int scanInterval)
 
     public void setScanDirs(List<File> dirs)
     {
-        _scanDirs.clear();
-        _scanDirs.addAll(dirs);
+        _scannables.clear();
+        if (dirs == null)
+            return;
+
+        for (File f:dirs)
+        {
+            addScanDir(f);
+        }
     }
 
+    @Deprecated
     public synchronized void addScanDir(File dir)
     {
-        _scanDirs.add(dir);
+        if (dir == null)
+            return;
+        try
+        {
+
+            if (dir.isDirectory())
+                addDirectory(dir.toPath());
+            else
+                addFile(dir.toPath());
+        }
+        catch (Exception e)
+        {
+            LOG.warn(e);
+        }
+    }
+    
+    /**
+     * Add a file to be scanned. The file must not be null, and must exist.
+     * 
+     * @param p the Path of the file to scan.
+     * @throws IOException 
+     */
+    public synchronized void addFile(Path p) throws IOException
+    {
+        if (p == null)
+            throw new IllegalStateException("Null path");
+        
+        File f = p.toFile();
+        if (!f.exists() || f.isDirectory())
+            throw new IllegalStateException("Not file or doesn't exist: " + f.getCanonicalPath());
+        _scannables.put(p, null);
     }
 
+    /**
+     * Add a directory to be scanned. The directory must not be null and must exist.
+     * 
+     * @param p the directory to scan.
+     * @return an IncludeExcludeSet<PathMatcher,Path> to which the caller can add PathMatcher patterns to match
+     * @throws Exception
+     */
+    public synchronized IncludeExcludeSet<PathMatcher, Path> addDirectory(Path p)
```

Comment:
perhaps call this `scanDirectory` as it may not add if it is already there!

Reply:
I don't want to call the method scanDirectory as it's a verb and to me implies that the scan would be done by the method rather than configuring a directory to be scanned later.